File: c#/01-SingleResponsibilityPrinciple.cs

// Single Responsibility Principle Violation

class Report
{
    public string GetTitle()
    {
        return "Report Title";
    }

    public string GetDate()
    {
        return "2016-04-21";
    }

    public object GetContents()
    {
        return new {
            Title = GetTitle(),
            Date = GetDate()
        };
    }

    public string FormatJson()
    {
        return JsonSerializer.Serialize(GetContents());
    }
}

// Refactored

class Report
{
    public string GetTitle()
    {
        return "Report Title";
    }

    public string GetDate()
    {
        return "2016-04-21";
    }

    public object GetContents()
    {
        return new {
            Title = GetTitle(),
            Date = GetDate()
        };
    }
}

interface ReportFormattable
{
    Format(Report report);
}

class JsonReportFormatter : ReportFormattable
{
    public Format(Report report)
    {
        return JsonSerializer.Serialize(GetContents());
    }
}

File: js/01-singleResponsibilityPrinciple.js

// Single Responsibility Principle Violation

class Report {

    getTitle() {
        return 'Report Title';
    }

    getDate() {
        return '2016-04-21';
    }

    getContents() {
        return {
            title: this.getTitle(),
            date: this.getDate()
        };
    }

    formatJson() {
        return JSON.stringify(this.getContents());
    }
}

// Refactored

class Report {

    getTitle() {
        return 'Report Title';
    }

    getDate() {
        return '2016-04-21';
    }

    getContents() {
        return {
            title: this.getTitle(),
            date: this.getDate()
        };
    }
}

class JsonReportFormatter {

    format(report) {
        return JSON.stringify(report.getContents());
    }
}
